File: frontend/src/api/leagues.ts
// Query/mutation hooks for /api/leagues — Settings' "ESPN Leagues" card (task 7.3).
// GET returns a plain list across both platforms (no envelope — this is Settings
// config data, not a Dashboard read); EspnLeaguesCard filters to platform === "espn".

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import type { Platform, ScoringType } from "../types/api";
import { apiClient } from "./client";

export interface LeagueSetting {
  id: string;
  platform: Platform;
  platform_id: string;
  name: string;
  season: number;
  team_count: number;
  scoring_type: ScoringType;
  current_week: number;
  is_enabled: boolean;
}

export const LEAGUES_QUERY_KEY = ["leagues"] as const;

export function useLeagues() {
  return useQuery({
    queryKey: LEAGUES_QUERY_KEY,
    queryFn: () => apiClient.get<LeagueSetting[]>("/api/leagues"),
  });
}

export function useUpdateLeague() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ leagueId, isEnabled }: { leagueId: string; isEnabled: boolean }) =>
      apiClient.patch<LeagueSetting>(`/api/leagues/${leagueId}`, { is_enabled: isEnabled }),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: LEAGUES_QUERY_KEY });
      // A disabled league's teams drop out of GET /api/teams (backend/gridiron/
      // services/fantasy_service.py:list_teams) — keep the Dashboard in sync.
      void queryClient.invalidateQueries({ queryKey: ["teams"] });
    },
  });
}

File: backend/gridiron/api/leagues.py
"""`/api/leagues` — list every discovered league (both platforms) and toggle whether a
league counts toward the app (task 7.3, Settings' "ESPN Leagues" card).

Plain list response, not the `Envelope` (design.md D12) — this is Settings config data,
not a Dashboard read that needs `live_state`/`as_of`/`platforms` freshness metadata.
"""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from backend.gridiron import schemas
from backend.gridiron.db import get_session
from backend.gridiron.models import League

router = APIRouter(prefix="/api/leagues", tags=["leagues"])


class LeagueOut(BaseModel):
    id: str
    platform: schemas.Platform
    platform_id: str
    name: str
    season: int
    team_count: int
    scoring_type: schemas.ScoringType
    current_week: int
    is_enabled: bool


class LeagueUpdateRequest(BaseModel):
    is_enabled: bool


def _to_out(row: League) -> LeagueOut:
    return LeagueOut(
        id=row.id,
        platform=row.platform,
        platform_id=row.platform_id,
        name=row.name,
        season=row.season,
        team_count=row.team_count,
        scoring_type=row.scoring_type,
        current_week=row.current_week,
        is_enabled=row.is_enabled,
    )


@router.get("", response_model=list[LeagueOut])
async def list_leagues(session: AsyncSession = Depends(get_session)) -> list[LeagueOut]:
    rows = (
        (await session.execute(select(League).order_by(League.platform, League.name)))
        .scalars()
        .all()
    )
    return [_to_out(row) for row in rows]


@router.patch("/{league_id}", response_model=LeagueOut)
async def update_league(
    league_id: str,
    body: LeagueUpdateRequest,
    session: AsyncSession = Depends(get_session),
) -> LeagueOut:
    row = await session.get(League, league_id)
    if row is None:
        raise HTTPException(
            status_code=404,
            detail={"code": "league_not_found", "message": f"unknown league id: {league_id}"},
        )
    row.is_enabled = body.is_enabled
    await session.commit()
    await session.refresh(row)
    return _to_out(row)
